share/vss/vss.go

```diff
@@ -0,0 +1,630 @@
+		return errors.New("vss: invalid t received in Deal")
+	}
+
+	if !bytes.Equal(a.sid, d.SessionID) {
+		return errors.New("vss: find different sessionIDs from Deal")
+	}
+
+	if err := sign.VerifySchnorr(a.suite, a.dealer, msgDeal(d), d.Signature); err != nil {
+		return err
+	}
+
+	fi := d.SecShare
+	gi := d.RndShare
+	if fi.I != gi.I {
+		return errors.New("vss: not the same index for f and g share in Deal")
+	} else if fi.I < 0 || fi.I >= len(a.verifiers) {
+		return errors.New("vss: index out of bounds in Deal")
+	}
+	// compute fi * G + gi * H
+	fig := a.suite.Point().Base().Mul(nil, fi.V)
+	H := deriveH(a.suite, a.verifiers)
+	gih := a.suite.Point().Mul(H, gi.V)
+	ci := a.suite.Point().Add(fig, gih)
+
+	commitPoly := share.NewPubPoly(a.suite, nil, d.Commitments)
+
+	pubShare := commitPoly.Eval(fi.I)
+	if !ci.Equal(pubShare.V) {
+		return errors.New("vss: share do not verify against commitments in Deal")
+	}
+	return nil
+}
+
+func (a *aggregator) verifyResponse(r *Response) error {
+	if !bytes.Equal(r.SessionID, a.sid) {
+		return errors.New("vss: receiving inconsistent sessionID in response")
+	}
+
+	pub, ok := findPub(a.verifiers, r.Index)
+	if !ok {
+		return errors.New("vss: index out of bounds in response")
+	}
+
+	if err := sign.VerifySchnorr(a.suite, pub, msgResponse(r), r.Signature); err != nil {
+		return err
+	}
+
+	return a.addResponse(r)
+}
+
+func (a *aggregator) verifyJustification(j *Justification) error {
+	if _, ok := findPub(a.verifiers, j.Index); !ok {
+		return errors.New("vss: index out of bounds in justification")
+	}
+	r, ok := a.responses[j.Index]
+	if !ok {
+		return errors.New("vss: no complaints received for this justification")
+	} else if r.Status != StatusComplaint {
+		return errors.New("vss: justification received for an approval")
+	}
+
+	if err := a.VerifyDeal(j.Deal, false); err != nil {
+		// if one response is bad, flag the dealer as malicious
+		a.badDealer = true
+		return err
+	}
+	r.Status = statusJustified
+	return nil
+}
+
+func (a *aggregator) addResponse(r *Response) error {
+	if _, ok := findPub(a.verifiers, r.Index); !ok {
+		return errors.New("vss: index out of bounds in Complaint")
+	}
+	if _, ok := a.responses[r.Index]; ok {
+		return errors.New("vss: already existing response from same origin")
+	}
+	a.responses[r.Index] = r
+	return nil
+}
+
+// EnoughApprovals returns true if enough verifiers have sent their approval for
+// the deal they received.
+func (a *aggregator) EnoughApprovals() bool {
+	var app int
+	for _, r := range a.responses {
+		if r.Status == StatusApproval {
+			app++
+		}
+	}
+	return app >= a.t
+}
+
+// DealCertified returns true if there has been less than t complaints, all
+// Justifications were correct and if EnoughApprovals() returns true.
+func (a *aggregator) DealCertified() bool {
+	var comps int
+	for _, r := range a.responses {
+		if r.Status == StatusComplaint {
+			comps++
+		}
+	}
+	return a.EnoughApprovals() && !(comps >= a.t || a.badDealer)
```

Comment:
style: as this is not `C`, I would put the `!(comps...)` into its own variable and give it a meaningful name.